Add Modernizing Kubeflow Pipelines UI blog post (#196)
kubeflow/blog: _posts/2026-03-31-modernizing-kubeflow-pipelines-ui.md

---
toc: true
layout: post
comments: true
title: "Modernizing Kubeflow Pipelines UI"
hide: false
categories: [pipelines]
permalink: /modernizing-kubeflow-pipelines-ui/
author: Manaswini Das
---

> The Kubeflow Pipelines web interface has been upgraded from React 16 to React 19 — a modernization effort that touches every layer of the frontend stack. Whether you use the UI to manage pipelines day-to-day or contribute to the codebase, here's what this means for you.

Reviewer: ```suggestion
The Kubeflow Pipelines web interface has been upgraded from React 16 to React 19 — a modernization effort that touches every layer of the frontend stack. Whether you use the UI to manage pipelines day-to-day or contribute to the codebase, here is what this means for you.
```

## What's changing for users

> You don't need to do anything differently. Your bookmarks, workflows, and browser all work exactly as before. But under the hood, the UI is now built on a modern foundation that delivers tangible improvements:

Reviewer: ```suggestion
You do not need to do anything differently. Your bookmarks, workflows, and browser all work exactly as before. But under the hood, the UI is now built on a modern foundation that delivers tangible improvements:
```

### Improved charts and metrics display

Run metrics and comparison charts now use Recharts instead of the deprecated react-vis library. The new charting library renders more efficiently, handles edge cases better, and provides cleaner visual output when comparing run results side by side.

### Better accessibility

The component library migration from Material-UI v3 to MUI v5 brings improved keyboard navigation, better ARIA attribute coverage, and more consistent focus management across dialogs, tables, and form elements. These improvements make the UI more usable with screen readers and keyboard-only workflows.

### No breaking changes

Every user-facing feature works the same way it did before. The API contracts are unchanged. If you use the KFP Python SDK or REST API to interact with the platform, nothing changes on your end. This upgrade was purely a frontend modernization — zero impact on backend behavior, pipeline execution, or artifact storage.

## Why we made this change

> The KFP frontend had been running on React 16 (released in 2017) with Material-UI v3, Create React App, and Jest/Enzyme for testing. This created compounding issues:

Reviewer: ```suggestion
The KFP frontend had been running on React 16 (released in 2017) with Material-UI v3, create-react-app, and Jest/Enzyme for testing. This created compounding issues:
```

Author: https://github.com/kubeflow/blog/pull/196#discussion_r3016070785

Reviewer: i updated it now to make clear what youare referencing so create-react-app

- **Security exposure.** React 16 and 17 no longer receive security patches, and dozens of transitive dependencies were locked to outdated versions because of React peer constraints.

> - **Stalled ecosystem.** Modern libraries — including improved data-fetching, visualization, and accessibility tools — dropped support for React 16/17. Staying behind meant the UI couldn't benefit from upstream improvements.

Reviewer: ```suggestion
- **Stalled ecosystem.** Modern libraries — including improved data-fetching, visualization, and accessibility tools — dropped support for React 16/17. Staying behind meant the UI could not benefit from upstream improvements.
```

- **Contributor friction.** The legacy CRA + Jest + Enzyme toolchain was slow to build, brittle to test, and increasingly difficult for new contributors to set up. Modernizing the stack lowers the barrier to contribution.

## How we got here

Rather than attempting a single risky version jump, we followed a deps-first, bump-last strategy: upgrade every dependency to be forward-compatible before touching React itself. A custom React peer compatibility gate in CI prevented regressions at every step. The work was executed across **20+** pull requests in strict dependency order.

### React 16 → 17: Rebuilding the foundation

> Before React could move forward, the entire build and test toolchain had to be replaced. Create React App was swapped for Vite, Jest + Enzyme gave way to Vitest + Testing Library, and Material-UI was upgraded from v3 to v4 to unblock the React 17 peer range. The deprecated react-vis charting library was replaced with Recharts. With those blockers cleared, the React 17 bump itself was a small, low-risk change.

Reviewer: ```suggestion
Before React could move forward, the entire build and test toolchain had to be replaced. create-react-app was swapped for Vite, Jest + Enzyme gave way to Vitest + Testing Library, and Material-UI was upgraded from v3 to v4 to unblock the React 17 peer range. The deprecated react-vis charting library was replaced with Recharts. With those blockers cleared, the React 17 bump itself was a small, low-risk change.
```

Author: This is pertaining to a proper noun - can't we retain this ? https://www.npmjs.com/package/create-react-app

Reviewer: i updated it now to make clear what youare referencing so create-react-app
